### ISSUE TITLE
FacetGrid with independent colorbars

### ISSUE THREAD

**Issue by nbrenow** [pydata/xarray#1387]
Sometimes the magnitude of a variable can vary dramatically across a given coordinate, which makes 2d plots generated by xr.FacetGrid difficult to interpret. It would be useful if an option to xr.FacetGrid could be specified which allows each subplot to have its own colorbar.

**Comment by fmaussion** [comment]
Seems reasonable, although this would imply quite a lot of change in the current logic. For example, facetgrids share x and/or y axis:

![index](https://cloud.githubusercontent.com/assets/10050469/25519732/590ef63e-2bf9-11e7-87f0-cf3a3f072c44.png)

If there is a colorbar in between, would you like to share axes?

The current way to do what you need is to do it manually:

```
ds = xr.tutorial.load_dataset('air_temperature')
t = ds.air.groupby('time.season').mean(dim='time')
f, (ax1, ax2) = plt.subplots(1, 2, figsize=(7, 3))
t.isel(season=0).plot(ax=ax1)
t.isel(season=1).plot(ax=ax2)
```

![index2](https://cloud.githubusercontent.com/assets/10050469/25519839/bc461cfa-2bf9-11e7-9736-fa6ae785fab6.png)

As you see, this messes up with the layout so a call to ``tight_layout()`` is necessary.

**Comment by nbrenow** [comment]
whoops. accidentally hit the close button.

@fmaussion I was worried that would be the case. I think matplotlib's [AxesGrid](https://matplotlib.org/2.0.0/mpl_toolkits/axes_grid/index.html) module is the appropriate tool to handle this use case. Unfortunately, I find the AxesGrid API pretty unpleasant for interactive use.

**Comment by nbrenow** [comment]
The AxesGrid API isn't actually that bad. See this

```python
from mpl_toolkits.axes_grid1 import ImageGrid

import xarray as xr

fig = plt.figure(figsize=(6,2.5))
grid = ImageGrid(fig, 111, (1,2),
                 axes_pad=.1,
                 cbar_location='bottom',
                 cbar_mode='each',
                 cbar_pad=0.3,
                 aspect=False,
                 cbar_size='3%')

for (lab,data), ax, cax in zip(t.groupby('season'), grid, grid.cbar_axes):
    im = ax.contourf(data)
    cax.colorbar(im)
```
![](http://i.imgur.com/HroE4za.png)

**Comment by shoyer** [comment]
This does kind of go against the point of FacetGrid, which is to display facets of the same variables in a grid. If your plots should have different colorbars, then arguably they should be different data variables in a Dataset.

That said, this does seem similar in spirit to the `sharex` and `sharey` arguments. Maybe `sharez` or `share_vlim` would be appropriate arguments for changing this?

Squeezing multiple colorbars in will be a little trickier than just allowing colorbar limits to vary, but if it can be done in a sane fashion that doesn't make the existing code more complex I would be in support.

**Comment by nbrenow** [comment]
I tend to agree with you @shoyer. In that case it would be nice to have a `Dataset.plot` convenience method for plotting all the variables in a Dataset. This method could handle all the annoying boiler plate associated with subplots/AxesGrid.

**Comment by pollackscience** [comment]
I like @shoyer 's suggestion as well.  This would be very useful if the data being plotted are equal-sized images, in which case the colorbar could be dropped altogether as it no longer makes sense for them to share the same range.

**Comment by tom-andersson** [comment]
Any updates on this? It would be great to have an option for multiple colorbars for each axis of the `FacetGrid`, rather than the default of one single colorbar for each axis. The main use case would be when different axes span different orders of magnitude. A global `clim` hides the variability of the smaller-scale axes.

@shoyer:
> If your plots should have different colorbars, then arguably they should be different data variables in a Dataset.

I see the reasoning here. Could the solution be an `xr.Dataset.plot.imshow` method, which optionally takes a list of `cmap`s, that automatically produces a new colorbar for each `data_var`/axis?

In the meantime, manual subplot and colorbar construction with `matplotlib` works, but is annoying.

**Comment by etienneschalk** [comment]
Hello, I came from https://stackoverflow.com/questions/70667835/use-different-color-scales-in-xarray-faceted-imshow

And when working with for instance multi-band GeoTiff containing different variables, having different colorscales would be nice

**Comment by dcherian** [comment]
This may be cleaner to do with a Dataset FacetGrid plot, where one specifies different arrays across rows or columns. and colorbars are only shared per array. Doesn't directly solve the OP's problem, but adding yet another kwarg doesn't seem like a good option.
